### PR TITLE
[FLINK-16705] Ensure MiniCluster shutdown does not interfere with JobResult retrieval

### PR DESCRIPTION
**Pull request by mxm**
## What is the purpose of the change

There is a race condition in `LocalExecutor` between (a) shutting down the
cluster when the job has finished and (b) the client which retrieves the result
of the job execution.

This was observed in Beam, running a large test suite with the Flink Runner.

## Brief change log

We should make sure the job result retrieval and the cluster shutdown do not
interfere. This adds a PerJobMiniCluster and a corresponding JobClient which guarantee that.

## Verifying this change

This change added tests, see `PerJobMiniClusterTest` and `LocalExecutorITCase`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (JavaDocs)

### CODE REVIEW

**Comment by mxm** [comment]
@flinkbot run travis
@flinkbot run azure

**Comment by mxm** [comment]
@flinkbot run travis

**Comment by mxm** [comment]
@flinkbot run azure

File: flink-clients/src/test/java/org/apache/flink/client/program/PerJobMiniClusterClientTest.java [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+/**
+ * Tests for {@link PerJobMiniClusterClient}.
+ */
+public class PerJobMiniClusterClientTest {
```
Comment (tillrohrmann) [inline flink-clients/src/test/java/org/apache/flink/client/program/PerJobMiniClusterClientTest.java]:
```suggestion
public class PerJobMiniClusterClientTest extends TestLogger {
```

File: flink-tests/src/test/java/org/apache/flink/test/example/client/LocalExecutorITCase.java [hunk 2]
```diff
@@ -81,4 +114,35 @@ private Plan getWordCountPlan(File inFile, File outFile, int parallelism) {
...
+		@Override
+		protected MiniCluster startMiniCluster(JobGraph jobGraph, Configuration configuration) throws Exception {
+			return miniCluster = super.startMiniCluster(jobGraph, configuration);
+		}
+
+		public boolean isMiniClusterRunning() {
+			Preconditions.checkNotNull(miniCluster, "MiniCluster has not been started yet.");
+			return miniCluster.isRunning();
+		}
+	}
```
Comment (tillrohrmann) [inline flink-tests/src/test/java/org/apache/flink/test/example/client/LocalExecutorITCase.java]: I think the proper way of making the `LocalExecutor` inspectable is not to increase the visibility of methods such as `startMiniCluster` but by refactoring the `LocalExecutor` to take a `MiniClusterFactory`. That way you could control the `MiniCluster` instance which is created.

File: flink-clients/src/test/java/org/apache/flink/client/program/PerJobMiniClusterClientTest.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+		// Trigger fetching the JobResult
+		jobIdFuture.complete(jobGraph.getJobID());
+
+		if (withException) {
+			jobResult.completeExceptionally(new RuntimeException("oups"));
+		} else {
+			jobResult.complete(mock(JobResult.class));
+		}
+
+		verify(clusterClient).shutDownCluster();
```
Comment (tillrohrmann) [inline flink-clients/src/test/java/org/apache/flink/client/program/PerJobMiniClusterClientTest.java]: We should try to reduce the usage of Mockito as it makes testing code harder to refactor. I think the problem here is that we don't have a proper `MiniCluster` interface which we could instrument. Maybe we should introduce this instead of using Mockito.

File: flink-clients/src/main/java/org/apache/flink/client/program/PerJobMiniClusterClient.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+ * A Client to interact with a {@link MiniCluster} and manage its lifecycle for a single job.
+ *
+ * <p>Shuts down the MiniCluster after the JobResult has been received. It caches the JobResult to ensure the result can
+ * be retrieved even if the MiniCluster has been shut down.
+ *
+ * <p>We want to do this to:
+ * (a) ensure the MiniCluster is torn down after the job completes (to avoid resource leakage)
+ * (b) avoid race condition between the async shutdown code and the job result retrieval.
+ */
+public class PerJobMiniClusterClient extends MiniClusterClient {
```
Comment (tillrohrmann) [inline flink-clients/src/main/java/org/apache/flink/client/program/PerJobMiniClusterClient.java]: I think with this class we are breaking the Liskov substitution principle. A `PerJobMiniClusterClient` no longer respects the contract of the `MiniClusterClient` as we are only allowed to submit a single job. An indicator for this problem are the many check states in the methods. Hence, a `PerJobMiniClusterClient` should not be a `MiniClusterClient`. A `MiniClusterClient` could be a `PerJobMiniClusterClient`, though.

**Comment by mxm** [comment]
Thank you for your comments @tillrohrmann. I agree with your assessment. I had very similar thoughts when I implemented this. The decision I made was a tradeoff between fixing this critical path for test execution in time for the 1.10.1 release vs doing a larger refactor which might not be suitable for a patch release.

I'll incorporate your feedback and post an update as soon as possible.

File: flink-tests/src/test/java/org/apache/flink/test/example/client/LocalExecutorITCase.java (hunk 2 above)
Comment (mxm) [reply flink-tests/src/test/java/org/apache/flink/test/example/client/LocalExecutorITCase.java]: IMHO that's a personal preference, either way we alter the main code to make it testable. Introducing additional abstractions for testing can make the code harder to read and maintain. If there is a factory for everything, it can be harder to keep track of what the code is doing.

Nevertheless, I think your suggestion makes sense. I've implemented a factory in the latest commit.

File: flink-clients/src/test/java/org/apache/flink/client/program/PerJobMiniClusterClientTest.java (hunk 3 above)
Comment (mxm) [reply flink-clients/src/test/java/org/apache/flink/client/program/PerJobMiniClusterClientTest.java]: I've drastically reduced the usage of mocks, there is just one test case left.

File: flink-clients/src/main/java/org/apache/flink/client/program/PerJobMiniClusterClient.java (hunk 4 above)
Comment (mxm) [reply flink-clients/src/main/java/org/apache/flink/client/program/PerJobMiniClusterClient.java]: There is now a `PerJobMiniCluster`.

**Comment by mxm** [comment]
@tillrohrmann I've had another shot at it today. PTAL.

**Comment by mxm** [comment]
@flinkbot run azure

**Comment by mxm** [comment]
Travis passes, though the Flink bot doesn't seem to recognize that: https://travis-ci.com/github/flink-ci/flink/builds/155934908

File: flink-clients/src/main/java/org/apache/flink/client/program/PerJobMiniCluster.java [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+			.handle((result, throwable) -> {
+				if (throwable != null) {
+					try {
+						throw new RuntimeException("Error submitting job to MiniCluster.", throwable);
+					} finally {
+						shutDownCluster(miniCluster);
+					}
+				}
+				return new PerJobMiniClusterJobClient(result.getJobID(), miniCluster);
+			});
```
Comment (tillrohrmann) [inline flink-clients/src/main/java/org/apache/flink/client/program/PerJobMiniCluster.java]: I think it is a good practice to separate the happy path and error handling:
```
return miniCluster
		.submitJob(jobGraph)
		.thenApply(jobSubmissionResult -> (JobClient) new PerJobMiniClusterJobClient(jobSubmissionResult.getJobID(), miniCluster))
		.whenComplete((ignored, throwable) -> {
			if (throwable != null) {
				shutDownCluster(miniCluster);
			}
		});
```

File: flink-clients/src/main/java/org/apache/flink/client/program/PerJobMiniCluster.java [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+
+	private static void shutDownCluster(MiniCluster miniCluster) {
+		try {
+			miniCluster.closeAsync().get();
+		} catch (InterruptedException e) {
+			Thread.currentThread().interrupt();
+		} catch (ExecutionException e) {
+			LOG.error("Error while shutting down cluster", e);
+		}
+	}
```
Comment (tillrohrmann) [inline flink-clients/src/main/java/org/apache/flink/client/program/PerJobMiniCluster.java]:
```suggestion
	private static void shutDownCluster(MiniCluster miniCluster) {
		miniCluster.closeAsync().whenComplete((ignored, throwable) -> {
			if (throwable != null) {
				LOG.warn("Shut down of MiniCluster failed.", throwable);
			}
		});
	}
```

File: flink-clients/src/test/java/org/apache/flink/client/program/PerJobMiniClusterTest.java [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+		private final Object lock = new Object();
+
+		public MyCancellableInvokable(Environment environment) {
+			super(environment);
+		}
+
+		@Override
+		public void invoke() throws Exception {
+			synchronized (lock) {
+				lock.wait();
```
Comment (tillrohrmann) [inline flink-clients/src/test/java/org/apache/flink/client/program/PerJobMiniClusterTest.java]: According to the JavaDocs of `wait`

>As in the one argument version, interrupts and spurious wakeups are possible, and this method should always be used in a loop:

Hence I would suggest to harden this usage.

File: flink-clients/src/test/java/org/apache/flink/client/program/PerJobMiniClusterTest.java [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+			.submitJob(cancellableJobGraph)
+			.get();
+
+		assertThat(jobClient.getJobID(), is(cancellableJobGraph.getJobID()));
+		assertThat(jobClient.getJobStatus().get(), is(JobStatus.RUNNING));
+
+		jobClient.cancel().get();
+
+		try {
+			jobClient.getJobExecutionResult(getClass().getClassLoader()).get();
```
Comment (tillrohrmann) [inline flink-clients/src/test/java/org/apache/flink/client/program/PerJobMiniClusterTest.java]: Are we expecting a certain exception here or that his method passes?

File: flink-clients/src/test/java/org/apache/flink/client/program/PerJobMiniClusterTest.java [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+
+	@Test
+	public void testSubmissionError() throws Exception {
+		PerJobMiniCluster perJobMiniCluster = initializeMiniCluster();
+
+		// JobGraph is not a valid job
+		JobGraph jobGraph = new JobGraph();
+
+		try {
+			perJobMiniCluster.submitJob(jobGraph).get();
```
Comment (tillrohrmann) [inline flink-clients/src/test/java/org/apache/flink/client/program/PerJobMiniClusterTest.java]: `fail(...)` missing.

File: flink-clients/src/main/java/org/apache/flink/client/program/PerJobMiniCluster.java (hunk 6 above)
Comment (tillrohrmann) [reply flink-clients/src/main/java/org/apache/flink/client/program/PerJobMiniCluster.java]: If the contract of `shutDownCluster` is that `miniCluster` is shut down after this method leaves, then my suggestion won't work.

File: flink-clients/src/test/java/org/apache/flink/client/program/PerJobMiniClusterTest.java [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+
+	@Test
+	public void testJobClientInteractionAfterShutdown() throws Exception {
+		PerJobMiniCluster perJobMiniCluster = initializeMiniCluster();
+		JobClient jobClient = perJobMiniCluster.submitJob(getNoopJobGraph()).get();
+		jobClient.getJobExecutionResult(getClass().getClassLoader()).get();
+		assertThatMiniClusterIsShutdown();
+
+		try {
+			jobClient.cancel();
```
Comment (tillrohrmann) [inline flink-clients/src/test/java/org/apache/flink/client/program/PerJobMiniClusterTest.java]: `fail` is missing.

File: flink-clients/src/main/java/org/apache/flink/client/program/PerJobMiniCluster.java [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+	/**
+	 * A {@link JobClient} for a {@link PerJobMiniCluster}.
+	 */
+	private static class PerJobMiniClusterJobClient implements JobClient {
+
+		private final JobID jobID;
+		private final MiniCluster miniCluster;
+		private final CompletableFuture<JobResult> jobResultFuture;
+
+		public PerJobMiniClusterJobClient(JobID jobID, MiniCluster miniCluster) {
```
Comment (tillrohrmann) [inline flink-clients/src/main/java/org/apache/flink/client/program/PerJobMiniCluster.java]: Public constructor in a non-public class.

File: flink-clients/src/test/java/org/apache/flink/client/program/PerJobMiniClusterTest.java [hunk 12]
```diff
@@ -0,0 +1,234 @@
...
+		});
+
+		JobClient jobClient = perJobMiniCluster.submitJob(jobGraph).get();
+
+		jobClient.triggerSavepoint(savepointPath);
+		Mockito.verify(miniCluster).triggerSavepoint(jobGraph.getJobID(), savepointPath, false);
+
+		jobClient.stopWithSavepoint(true, savepointPath);
+		Mockito.verify(miniCluster).stopWithSavepoint(jobGraph.getJobID(), savepointPath, true);
+	}
```
Comment (tillrohrmann) [inline flink-clients/src/test/java/org/apache/flink/client/program/PerJobMiniClusterTest.java]: I'm not so sure about the value of this test case. We setup quite some Mockito machinery to test that `PerJobMiniClusterJobClient` forwards `triggerSavepoint` to `MiniCluster`. I think this does not justify the complexity of this test. Since we don't test `stopWithSavepoint`, I would be fine with removing this test case given that its test scope is quite limited anyway.

Maybe it would be easier to test this aspect if we introduced a new interface for representing the interaction between the `MiniCluster` and the `PerJobMiniClusterJobClient`.

File: flink-clients/src/main/java/org/apache/flink/client/program/PerJobMiniCluster.java [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+			.setConfiguration(configuration)
+			.setNumTaskManagers(numTaskManagers)
+			.setRpcServiceSharing(RpcServiceSharing.SHARED)
+			.setNumSlotsPerTaskManager(numSlotsPerTaskManager)
+			.build();
+	}
+
+	private static void shutDownCluster(MiniCluster miniCluster) {
+		try {
+			miniCluster.closeAsync().get();
```
Comment (tillrohrmann) [inline flink-clients/src/main/java/org/apache/flink/client/program/PerJobMiniCluster.java]: Blocking calls should be avoided if possible. The problem is that you have to make sure that this method won't be called from a `Thread` belonging to `miniCluster`. Otherwise this can deadlock because the `ThreadPool` cannot shut down if one of its thread is waiting here.

File: flink-clients/src/main/java/org/apache/flink/client/program/PerJobMiniCluster.java [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+		} catch (InterruptedException e) {
+			Thread.currentThread().interrupt();
+		} catch (ExecutionException e) {
+			LOG.error("Error while shutting down cluster", e);
```
Comment (tillrohrmann) [inline flink-clients/src/main/java/org/apache/flink/client/program/PerJobMiniCluster.java]: I don't really like that we only log problems with the shut down here instead of properly treating errors during shut down. I guess the underlying problem is that `JobClient` has not really an owning component which can make sure that everything has cleanly shut down at the end. This might be something to reconsider. Out of this scope, though.

File: flink-clients/src/main/java/org/apache/flink/client/program/PerJobMiniCluster.java [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+					} finally {
+						shutDownCluster(miniCluster);
+					}
+				}
+				return new PerJobMiniClusterJobClient(result.getJobID(), miniCluster);
+			});
+	}
+
+	private MiniClusterConfiguration getMiniClusterConfig(int maximumParallelism) {
+		Configuration configuration = this.configuration.clone();
```
[the remark this answers is not in the document]
Comment (mxm) [reply flink-clients/src/main/java/org/apache/flink/client/program/PerJobMiniCluster.java]: I'm curious, why is that a bad pattern? Because of a potential ambiguity with inheritance?

File: flink-clients/src/main/java/org/apache/flink/client/program/PerJobMiniCluster.java (hunk 6 above)
Comment (mxm) [reply flink-clients/src/main/java/org/apache/flink/client/program/PerJobMiniCluster.java]: I'd keep the existing behavior for ensure that the cluster shuts down. As you pointed out, there is no code in JobClient which handles this.

File: flink-clients/src/test/java/org/apache/flink/client/program/PerJobMiniClusterTest.java (hunk 10 above)
Comment (mxm) [reply flink-clients/src/test/java/org/apache/flink/client/program/PerJobMiniClusterTest.java]: Thanks for catching these. I'm used to using `assertThrows` which is available in more recent versions of Junit. I've added a method similar to this.

File: flink-clients/src/test/java/org/apache/flink/client/program/PerJobMiniClusterTest.java (hunk 12 above)
Comment (mxm) [reply flink-clients/src/test/java/org/apache/flink/client/program/PerJobMiniClusterTest.java]: I've removed the Mockito machinery (love that term). I've replaced it with a simple test that checks whether the savepoint is triggered.

File: flink-clients/src/main/java/org/apache/flink/client/program/PerJobMiniCluster.java (hunk 6 above)
Comment (mxm) [reply flink-clients/src/main/java/org/apache/flink/client/program/PerJobMiniCluster.java]: On the other hand, I think this is an improvement, so I'll adapt your suggestion.

**Comment by mxm** [comment]
Thanks a lot for the thorough review @tillrohrmann. Great suggestions. I've incorporated all of them. Please have another look.

File: flink-clients/src/test/java/org/apache/flink/client/program/PerJobMiniClusterTest.java (hunk 10 above)
Comment (tillrohrmann) [reply flink-clients/src/test/java/org/apache/flink/client/program/PerJobMiniClusterTest.java]: Yeah, we should bump our Junit dependency.

File: flink-clients/src/test/java/org/apache/flink/client/program/PerJobMiniClusterTest.java (hunk 12 above)
Comment (tillrohrmann) [reply flink-clients/src/test/java/org/apache/flink/client/program/PerJobMiniClusterTest.java]: Nice :-)

**Comment by mxm** [comment]
Thanks @tillrohrmann for working with me on this. Looks like the tests are green now.

File: flink-core/src/main/java/org/apache/flink/core/execution/PipelineExecutor.java [hunk 16]
```diff
@@ -41,5 +41,5 @@
 	 * @param configuration the {@link Configuration} with the required execution parameters
 	 * @return a {@link CompletableFuture} with the {@link JobClient} corresponding to the pipeline.
 	 */
-	CompletableFuture<JobClient> execute(final Pipeline pipeline, final Configuration configuration) throws Exception;
+	CompletableFuture<? extends JobClient> execute(final Pipeline pipeline, final Configuration configuration) throws Exception;
```
Comment (aljoscha) [inline flink-core/src/main/java/org/apache/flink/core/execution/PipelineExecutor.java:44]: @mxm Why was this changed from `CompletableFuture<? extends JobClient>`

Comment (mxm) [reply flink-core/src/main/java/org/apache/flink/core/execution/PipelineExecutor.java:44]: This was necessary to please Java generics when handling the derived JobClient class via futures. The alternative would have been to cast which we wanted to avoid. Since the class is marked `@Internal`, this should not be an issue?

Comment (aljoscha) [reply flink-core/src/main/java/org/apache/flink/core/execution/PipelineExecutor.java:44]: Yes, it's not a problem for the API. It just makes the internal signatures more wordy and I was surprised because I had to rebase some of my changes. 😅

I think the problem here is that Java in some situations can't type-check the code, if the `whenComplete()` would be removed it's happy to accept the result as a `CompletableFuture<JobClient>`.

Comment (mxm) [reply flink-core/src/main/java/org/apache/flink/core/execution/PipelineExecutor.java:44]: True, it makes things more verbose. If you find a way to remove it, go ahead. Perhaps sync with @tillrohrmann since he also reviewed the code and might have an idea.

Comment (tillrohrmann) [reply flink-core/src/main/java/org/apache/flink/core/execution/PipelineExecutor.java:44]: I think this change is not necessary and can be reverted. No explicit cast should be required if simply using an `thenApply(Function.identity())` with the right generics.
